### PR TITLE
Add support for user and shared vault type

### PR DESCRIPTION
**Pull request by tiran**
Signed-off-by: Christian Heimes <cheimes@redhat.com>

### CODE REVIEW

File: custodia/ipa/vault.py [hunk 1]
```diff
@@ -71,26 +70,54 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+        elif self.vault_type == 'user':
+            if self.user is None:
+                msg = "{!r}: User vault requires 'user' parameter"
+                raise ValueError(msg.format(self))
+            if six.PY2 and isinstance(self.user, str):
+                self.user = self.user.decode('utf-8')
+            self._vault_args = {'username': self.user}
+        elif self.vault_type == 'service':
+            if self.principal is None:
+                msg = "{!r}: User vault requires 'principal' parameter"
```
Comment (simo5) [inline custodia/ipa/vault.py]: Do we really require a principal ?
If you get the uid of the caller you can simply figure out the principal in IPA ?

Comment (simo5) [reply custodia/ipa/vault.py]: Err sorry I mean if we have a keytab we can simply get the principal from there, or not care and just let gssapi initiate the call with whatever principal is in the keytab

File: custodia/ipa/vault.py [hunk 2]
```diff
@@ -71,26 +70,54 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
     ipa_debug = PluginOption(bool, False, "debug mode for ipalib")
 
     def __init__(self, config, section=None):
         super(IPAVault, self).__init__(config, section)
+
+        if self.vault_type == 'shared':
+            self._vault_args = {'shared': True}
+        elif self.vault_type == 'user':
+            if self.user is None:
+                msg = "{!r}: User vault requires 'user' parameter"
```
Comment (simo5) [inline custodia/ipa/vault.py]: Here I wonder if we can get the UID with get_peer_creds and use that instead of hardcoding.

**Review by simo5 — COMMENTED** [review]
Any good reason to harcode values in configs ?

File: custodia/ipa/vault.py (hunk 1 above)
Comment (tiran) [reply custodia/ipa/vault.py]: I think FreeIPA wants a principal name for the API call. We'd have to get the principal from our keytab and then pass it along.

File: custodia/ipa/vault.py (hunk 2 above)
Comment (tiran) [reply custodia/ipa/vault.py]: interesting idea! How are we going to solve delegation?

File: custodia/ipa/vault.py (hunk 1 above)
Comment (simo5) [reply custodia/ipa/vault.py]: yes we can use python-gssapi to acquire creds from the ketab, then inquire creds to find the principal, you can then use it as is or parse/split it as needed.

File: custodia/ipa/vault.py (hunk 2 above)
Comment (simo5) [reply custodia/ipa/vault.py]: No idea about delegation, will we have information about it in IPA? Or will it be ad hoc configuration on the client side ?

**Review by simo5 — APPROVED** [review]
LGTM

**Pull request merged**